### PR TITLE
ArgoCD upgrade&cleanup

### PR DESCRIPTION
**Pull request by jkosik**
1. Upgrading ArgoCD to the latest version
2. Fixing incomplete cleanup (ArgoCD Custom Resources)
3. (optional) Conditional deletion of "argocd" namespace. Preventing accidental removal of all other Argo components sharing the same namespace. Oftentimes e.g. ArgoCD Image Updater or Argo Notifications share the same namespace with ArgoCD. Current version of deletion script may destroy complete Argo stack (regardless whether the other components are deployed as Addons or manually). Copy-pasting of this logic to further Argo Addons can even increase the danger.
4. ~~Fixes: #50~~ 

---

**Small glitch:**
- When microk8s has added community repo and my forked repo in parallel,
$ microk8s addons repo list
REPO       ADDONS SOURCE
core           17 (built-in)
**mycommunity     17 https://github.com/jkosik/microk8s-community-addons@0f1081**
**community      17 /snap/microk8s/current/addons/community@854c00**

- installation of my fork: 
$ microk8s enable mycommunity/argocd

- results in false reporting of installed Addons:
ubuntu@master:~$ microk8s status
...snipped...
addons:
  enabled:
     **argocd               # (community) Argo CD is a declarative continuous deployment for Kubernetes.**
     community            # (core) The community addons repository
     dns                  # (core) CoreDNS
     ha-cluster           # (core) Configure high availability on the current node
     helm3                # (core) Helm 3 - Kubernetes package manager
     **argocd               # (mycommunity) Argo CD is a declarative continuous deployment for Kubernetes.**

Seems as just reporting issue with colliding Addon names, since **forked** Addon was properly installed. Disabling Addon removes both clones from the list.
Worth issuing a bug?

BTW, great job with Microk8s!!!

### CODE REVIEW

**Comment by madalinignisca** [comment]
This NFS addon seems to support storage from one node. This is not acceptable on HA setup. The OpenEBS addon if it will be refactored should already provide NFS with HA out of the box.

**Comment by jkosik** [comment]
@madalinignisca, right, by default storage from one Node. It addresses current limitations of `hostpath-storage` and HA/anti-affinities on Pod level, not HA on Storage level. So not competing with OpenEBS, just different use-case and works already.
However (with some added complexity) NFS Provisioner can be hooked to e.g. external EBS, and get HA on Storage level as well (and also avoiding expensive EFS). It is more a question, how rich the Addon portfolio is planned to be.

**Comment by madalinignisca** [comment]
Should be well documented that by default will not offer HA and people hopping for an out of the box solution should try the alternative like OpenEBS. I will make a pool request with improved OpenEBS, including OpenEBS NFS component.

**Comment by madalinignisca** [comment]
Finished the new install. Need to do a better uninstall, exiting if there are pv(c)s in use.

I am interested also in this separate nfs provisioner. With hostpath could help some scenarios where HA is not required, but lower ram usage is critical (rpi4?)

OpenEBS with jiva and nfs adds 400mb from start on each node. Acceptable for my work.

I wanna get Longhorn also working. Seems since MicroK8S 1.22 is not working anymore.

**Comment by jkosik** [comment]
@neoaggelos sure, PR split was planned, my bad. I will do so asap. 
Ad `nfs-csi-provider`, correct "This driver requires existing and already configured NFSv3 or NFSv4 server". For NFS Server Provisioner there is no prereq...as said before, just another use-case, different pros/cons.

**Comment by jkosik** [comment]
@neoaggelos, PR updated, currently deals only with ArgoCD. Btw, Issue on NFS already exists #50

File: addons/argocd/disable [hunk 1]
```diff
@@ -15,9 +15,18 @@ echo "Disabling ArgoCD"
 $HELM delete argo-cd -n $NAMESPACE_ARGOCD
 
 $KUBECTL delete customresourcedefinition applications.argoproj.io \
+                applicationsets.argoproj.io \
                 argocdextensions.argoproj.io  \
                 appprojects.argoproj.io
 
-$KUBECTL delete $KUBECTL_DELETE_ARGS namespace "$NAMESPACE_ARGOCD" > /dev/null 2>&1 || true
+if [[ "$1" == "--purge" ]]; then
```
Comment (neoaggelos) [inline addons/argocd/disable:22]: This looks very restrictive, if you are adding arguments to the script, maybe make sure that the command line is parsed properly, see for example https://github.com/canonical/microk8s-core-addons/blob/a47841e2292ca6a82c8a98057010630c4865cd63/addons/ha-cluster/disable#L13-L27

**Review by neoaggelos — APPROVED** [review]
Thanks @jkosik

File: addons/argocd/disable [hunk 2]
```diff
@@ -15,9 +15,36 @@ echo "Disabling ArgoCD"
...
+          *) echo "Usage: microk8s disable argocd [-p]"
+             echo ""
+             echo "WARNING: Final deletion of \"$NAMESPACE_ARGOCD\" namespace must be enforced by: \"microk8s disable argocd -p\""
+             echo ""
+             echo "Purge only when sure, that \"$NAMESPACE_ARGOCD\" namespace is not hosting any other services from Argo stack."
+             echo ""             
+             exit 0
+             ;;
+  esac
+done
```
Comment (neoaggelos) [inline addons/argocd/disable:37]: Just noticed, this should be earlier in the script, before we remove the helm installation and the CRDs

**Review by neoaggelos — CHANGES_REQUESTED** [review]
Almost there, one final change. Thanks!

File: addons/argocd/disable (hunk 2 above)
Comment (jkosik) [reply addons/argocd/disable:28]: @neoaggelos right, makes sense. New commit added.

**Review by neoaggelos — APPROVED** [review]

**Pull request merged**